tools/c7n_azure/c7n_azure/resources/cosmos_db.py

```diff
@@ -403,6 +404,158 @@ def _query_rules(self, resource):
+
+        # We need to remove the 'magic string' they use for AzureCloud bypass
+        if '0.0.0.0' in parts:
+            parts.remove('0.0.0.0')
+
+        resource_rules = IPSet(filter(None, parts))
 
         return resource_rules
+
+
+@CosmosDB.action_registry.register('set-firewall-rules')
+class CosmosSetFirewallAction(SetFirewallAction):
+    """ Set Firewall Rules Action
+
+     Updates CosmosDB Firewall settings.  Learn about the firewall at:
+     https://docs.microsoft.com/en-us/azure/cosmos-db/firewall-support
+
+     By default the firewall rules are replaced with the new values.  The ``append``
+     flag can be used to force merging the new rules with the existing ones on
+     the resource.
+
+     You may also reference azure public cloud Service Tags by name in place of
+     an IP address.  Use ``ServiceTags.`` followed by the ``name`` of any group
+     from https://www.microsoft.com/en-us/download/details.aspx?id=56519.
+
+     Note that there are firewall rule number limits.  The limit for CosmosDB is
+     1000 rules (maximum tested rule count).
+
+     .. code-block:: yaml
+
+         - type: set-firewall-rules
+               ip-rules:
+                   - 11.12.13.0/16
+                   - ServiceTags.AppService.CentralUS
+
+
+     :example:
+
+     Find CosmosDB accounts without any firewall rules.
+
+     Enable the firewall and allow:
+     - All Azure Cloud IP space
+     - All Portal UI IP space
+     - Two additional external IP ranges
+
+     Mark ``append: True`` to ensure we only add to the existing configuration
+     which in this case means we don't remove any previously configured
+     vnet firewall rules.
+
+     .. code-block:: yaml
+
+        policies:
+          - name: cosmos-firewall
+            resource: azure.cosmosdb
+            filters:
+              # The firewall is disabled
+              - type: value
+                key: properties.ipRangeFilter
+                value: empty
+            actions:
+              - type: set-firewall-rules
+                append: True
+                bypass-rules:
+                  - AzureCloud
+                  - Portal
+                ip-rules:
+                  - 19.0.0.0/16
+                  - 20.0.1.2
+
+
+     Cosmos firewalls are disabled by simply configuring them with empty values.
+     We can do this with an empty action, which defaults to ``append: False``.
+
+     .. code-block:: yaml
+
+        policies:
+          - name: cosmos-firewall-clear
+            resource: azure.cosmosdb
+            filters:
+              # The firewall is enabled
+              - not:
+                - type: value
+                  key: properties.ipRangeFilter
+                  value: empty
+            actions:
+              - type: set-firewall-rules
+
+
+     """
+    schema = type_schema(
+        'set-firewall-rules',
+        required=[],
+        **{
+            'append': {'type': 'boolean', "default": False},
```

Comment:
nit: double quotes in `"default"` (and in some other places).. just to be consistent.